### ISSUE TITLE
Feature proposal: Ability to inline delegate parameters

### ISSUE THREAD

**Issue by Emik03** [oleg-st/InlineMethod.Fody#11]
It would be very nice (albeit difficult to implement) if you could add a way to annotate parameters that are delegates to be inlined whenever applicable.

Such a use case would look like this:

```cs
using System;

ReadOnlySpan<int> nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

Console.WriteLine(IndexOf(nums, x => x is 5));

[Inline]
static int IndexOf<T>(ReadOnlySpan<T> span, [Inline] Predicate<T> predicate)
{
    for (var i = 0; i < span.Length; i++)
        if (predicate(span[i]))
            return i;
    
    return -1;
}
```

...would compile to...

```cs

ReadOnlySpan<int> nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
int indexOf;

for (var i = 0; i < nums.Length; i++)
    if (nums[i] is 5)
    {
        indexOf = i;
        goto AfterIndexOf;
    }

indexOf = -1;

AfterIndexOf:

    Console.WriteLine(indexOf);
```

There are only two constraints that a delegate must adhere to in order to be inlined.
  - The delegate must only exist in the execution stack.
    - You must only use the `.Invoke` method within the method body.
    - You cannot cast it to its base types `Delegate` or `object`.
    - You cannot return the delegate, pass the delegate into a type's constructor.
    - You cannot pass the delegate as an argument into a function that doesn't also annotate [Inline] on that parameter.
  - The delegate must be statically resolved.
    - Its target must be `null`.
    - It cannot capture any variables.

I believe this can be detected by seeing if a display class is generated, *and* a cached field is exposed.

```cs
[Serializable]
[CompilerGenerated]
private sealed class <>c
{
    public static readonly <>c <>9 = new <>c();

    public static Predicate<int> <>9__0_0; // This line right here!

    internal bool <<Main>$>b__0_0(int x)
    {
        return x == 5;
    }
}
```

Keep in mind that the display class may reference an existing static function, in which case the class is static and therefore no instance is created, and the method body exists elsewhere:

```cs
[CompilerGenerated]
private static class <>O
{
    public static Predicate<int> <0>__IsFive;
}
```

Open questions:
- Should a warning or error be shown if you annotate `InlineAttribute` on a parameter that is not a delegate?
- Should we annotate using a different attribute instead of reusing the existing one?
- Should the display class be pruned?
- Should inlining occur across multiple variables, e.g.:

```cs
using System;

ReadOnlySpan<int> nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

ForEachComplex(
    nums,
    x => x % 2 is 0,
    x => Console.WriteLine($"Even: {x}"),
    x => Console.WriteLine($"Odd: {x}")
);

[Inline]
static void ForEachComplexBranching<T>(
    ReadOnlySpan<T> span,
    [Inline] Predicate<T> predicate,
    [Inline] Action<T> onTrue,
    [Inline] Action<T> onFalse
)
{
    for (var i = 0; i < span.Length; i++)
        if (predicate(span[i]))
            onTrue(span[i]);
        else
            onFalse(span[i]);
}
```

...can be optimized to...

```cs
using System;

ReadOnlySpan<int> nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

for (var i = 0; i < nums.Length; i++)
    if (nums[i] % 2 is 0)
        Console.WriteLine($"Even: {x}");
    else
        Console.WriteLine($"Odd: {x}");
```

**Comment by oleg-st** [comment]
Proposed only for delegates created inplace? IL pattern for delegate creation is not trivial, it contains branch.

`<>c.<>9__0_0 ?? (<>c.<>9__0_0 = new Predicate<int>(<>c.<>9.<Main>b__0_0))`

```csharp
    [Serializable]
    [CompilerGenerated]
    private sealed class <>c
    {
        public static readonly <>c <>9 = new <>c();

        public static Predicate<int> <>9__0_0;

        internal bool <Main>b__0_0(int x)
        {
            return x == 5;
        }
    }

    public static void Main(int t)
    {
        Console.WriteLine(IndexOf(RuntimeHelpers.CreateSpan<int>((RuntimeFieldHandle)/*OpCode not supported: LdMemberToken*/), <>c.<>9__0_0 ?? (<>c.<>9__0_0 = new Predicate<int>(<>c.<>9.<Main>b__0_0))));
    }
```

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhGAzAATZkDCZA3oWYxeRjkhSmQLICGAlgDsAFIIAuZUQEo6DJnIBKMHgBMA8gIA2ATwDKABx4CAPGIB8ZAQFcAtgGcyAXjIBtHABoyWDyQ8oPAVg8kDwB2DwAODwBODxwABgBdAG5ZOUZWKKEASQFlGARVADMhKzsPBEdzCr57f0lJFII5AF9UpjbWdjEyHLyC4sUVdW19QxMBUXNbAwEPAAVYZT4wHlEYccmyPUXl1ZhJNvomtMZC6DIhADceKDI+RzI4pLuyIzJpwwA6ABkYAQBzUQAC2efAA1GCDscTnI+IULtsYEsVmshB8BM4+Al6m0YWkMCE7o1cUwCWQ4DhGi1CM0gA=)

**Comment by Emik03** [comment]
That is true, but it appears in all cases of a branch it starts with an `ldsfld` and ends with a `stsfld` instruction, both to a field whose classes is auto-generated. You can delete that range of instructions which means you don't have to worry about the branch.

**Comment by oleg-st** [comment]
The first step can be to resolve the predicate call from the delegate argument:

```csharp
static int IndexOf(ReadOnlySpan<int> span, [ResolveDelegate] Predicate<int> predicate)
{
    for (var i = 0; i < span.Length; i++)
        if (predicate(span[i]))
            return i;

    return -1;
}
```

will became
```csharp
{
    for (var i = 0; i < span.Length; i++)
        if (<>c.<>9.<Main>b__0_0(span[i]))
            return i;

    return -1;
}
```

`<>c.<>9.<Main>b__0_0` can be inlined by JIT or InlineMethod if you add the Inline attribute to it:
```csharp
Console.WriteLine(IndexOf(nums, [Inline] (x) => x is 5));
```

**Comment by Emik03** [comment]
That'd be excellent! Is there any reason not to implicitly add the `InlineAttribute`, before passing it to the compile-time inliner? I think for most people if they want to inline the delegate indirection they also want to inline the method too. It's not a must, but I just think having to declare the delegate with `Inline` every time looks a bit noisy.

**Comment by oleg-st** [comment]
The first step is done

**Comment by oleg-st** [comment]
see https://github.com/oleg-st/InlineMethod.Fody/blob/master/src/InlineMethod.Tests.AssemblyToProcess/InlineDelegate.cs

`Caller` and `Inlined` will become the same.

**Comment by Emik03** [comment]
That is the exact type of feature I wanted to see, I'd be very glad to see this in the next update. Thanks in advance!

**Comment by oleg-st** [comment]
Released in 0.8
@Emik03 
Sorry for the delay, I was on vacation

**Comment by Emik03** [comment]
Unfortunately it seems to write invalid IL. It also failed to inline the delegate.

### Error
```
Unhandled exception. System.BadImageFormatException: An attempt was made to load a program with an incorrect format.
 (0x8007000B)
```
### Decompiled Source
```cs
    public Animations Change<T>(T value) where T : struct, Enum
    {
      T obj = value;
      Action<Animation<T>, M1> action = (Action<Animation<T>, M1>) ((a, v) => a.Change(v)); // No delegate should exist!
      M1 m1 = (M1) obj; // M1?
      Animations animations = this;
      Span<DrawableGameComponent> span = animations._animations.AsSpan<DrawableGameComponent>();
      ref DrawableGameComponent local1 = ref MemoryMarshal.GetReference<DrawableGameComponent>(span);
      for (ref DrawableGameComponent local2 = ref Unsafe.Add<DrawableGameComponent>(ref local1, span.Length); Unsafe.IsAddressLessThan<DrawableGameComponent>(ref local1, ref local2); local1 = ref Unsafe.Add<DrawableGameComponent>(ref local1, 1))
      {
        if (local1 is Animation<T> animation)
          ((Action<Animation<T>, T>) action)(animation, (T) m1);
      }
      return animations;
    }
```
### Original Source
```cs
    public Animations Change<T>(T value)
        where T : struct, Enum =>
        ForEach<T, T>(value, static (a, v) => a.Change(v));

    [Inline]
    Animations ForEach<T, TState>(
        TState state,
        [ResolveDelegate, RequireStaticDelegate(IsError = true)] Action<Animation<T>, TState> action
    )
        where T : struct, Enum
    {
        var animations = _animations.AsSpan();
        ref var start = ref MemoryMarshal.GetReference(animations);
        ref readonly var end = ref Unsafe.Add(ref start, animations.Length);

        for (; Unsafe.IsAddressLessThan(ref start, end); start = ref Unsafe.Add(ref start, 1))
            if (start is Animation<T> a)
                action(a, state);

        return this;
    }
```

**Comment by oleg-st** [comment]
1) Could you provide minimal repro code?
2) Possible bad IL is fixed in master

**Comment by Emik03** [comment]
Hi, sorry for a late response, here's an MRE. This definitely can get simplified further but in order for my use case to be fixed, you need to be able to resolve generics in the manner shown.
```cs
    [Inline]
    private bool Callee<T, TState>(
        TState state,
        [ResolveDelegate] Func<T, TState, bool> predicate
    )
    {
        return predicate(default(T), state);
    }

    public bool Caller<T>(int state)
    {
        return Callee<T, int>(state, (x, y) => (x.GetHashCode() + y) is 10);
    }

    public bool Inlined<T>(int state)
    {
        // I'm not sure if I'm supposed to include local variables, e.g.:
        // var x = default(T);
        // var y = state;
        // return (x.GetHashCode() + y) is 10;
        return (default(T).GetHashCode() + state) is 10;
    }
```
### Actual
```cs
 IL_0000: ldarg.0
 IL_0001: ldarg.1
 IL_0002: ldsfld System.Func`3<T,System.Int32,System.Boolean> InlineMethod.Tests.AssemblyToProcess.InlineDelegate/<>c__1`1<T>::<>9__1_0
 IL_0007: dup
 IL_0008: brtrue.s IL_0023
 IL_000a: pop
 IL_000b: ldsfld InlineMethod.Tests.AssemblyToProcess.InlineDelegate/<>c__1`1<T> InlineMethod.Tests.AssemblyToProcess.InlineDelegate/<>c__1`1<T>::<>9
 IL_0010: ldftn System.Boolean InlineMethod.Tests.AssemblyToProcess.InlineDelegate/<>c__1`1<T>::<Caller>b__1_0(T,System.Int32)
 IL_0016: newobj System.Void System.Func`3<T,System.Int32,System.Boolean>::.ctor(System.Object,System.IntPtr)
 IL_001b: dup
 IL_001c: stsfld System.Func`3<T,System.Int32,System.Boolean> InlineMethod.Tests.AssemblyToProcess.InlineDelegate/<>c__1`1<T>::<>9__1_0
 IL_0021: stloc.0
 IL_0022: ldloc.0
 IL_0023: stloc.2
 IL_0024: stloc.3
 IL_0025: pop
 IL_0026: ldloc.2
 IL_0027: ldloca.s V_1
 IL_0029: initobj T
 IL_002f: ldloc.1
 IL_0030: ldloc.3
 IL_0031: callvirt !2 System.Func`3<T,System.Int32,System.Boolean>::Invoke(!0,!1)
 IL_0036: ret
```
### Expected (no locals)
```cs
 IL_0000: ldloca.s V_0
 IL_0002: dup
 IL_0003: initobj T
 IL_0009: constrained. T
 IL_000f: callvirt System.Int32 System.Object::GetHashCode()
 IL_0014: ldarg.1
 IL_0015: add
 IL_0016: ldc.i4.s 10
 IL_0018: ceq
 IL_001a: ret
```
### Expected (with locals)
```cs
 IL_0000: ldloca.s V_0
 IL_0002: initobj T
 IL_0008: ldarg.1
 IL_0009: stloc.1
 IL_000a: ldloca.s V_0
 IL_000c: constrained. T
 IL_0012: callvirt System.Int32 System.Object::GetHashCode()
 IL_0017: ldloc.1
 IL_0018: add
 IL_0019: ldc.i4.s 10
 IL_001b: ceq
 IL_001d: ret
```

**Comment by Emik03** [comment]
Just to be clear, I have pulled upstream and it does not appear to pass the test either.

**Comment by oleg-st** [comment]
I made some fixes in master, see new tests

**Comment by Emik03** [comment]
Tests pass for me personally, and adding the assembly has produced valid IL. Thank you very much!
#### Decompiled source (now correct)
```cs
public Animations Change<T>(T value) where T : struct, Enum
{
    Span<DrawableGameComponent> span = this._animations.AsSpan<DrawableGameComponent>();
    ref DrawableGameComponent local1 = ref MemoryMarshal.GetReference<DrawableGameComponent>(span);
    for (ref DrawableGameComponent local2 = ref Unsafe.Add<DrawableGameComponent>(ref local1, span.Length); Unsafe.IsAddressLessThan<DrawableGameComponent>(ref local1, ref local2); local1 = ref Unsafe.Add<DrawableGameComponent>(ref local1, 1))
    {
        if (local1 is Animation<T> animation)
            animation.Change(value);
    }
    return this;
}
```

**Comment by oleg-st** [comment]
Released in 0.8.1

**Comment by oleg-st** [comment]
@Emik03 Can we close the issue as resolved?

**Comment by Emik03** [comment]
Sorry, I forgot to report one more bug:
```cs
class InlineDelegate3<TAnimation>
{
    public class Animation<T>
    {
        public void Change()
        {
        }
    }

    private Animation<TAnimation>[] _animations; 

    public void Caller<T>()
        where T : struct, Enum =>
        Callee(static (Animation<T> a) => a.Change());

    [Inline]
    void Callee<TAnimation>([ResolveDelegate] Action<TAnimation> action)
    {
        var animations = _animations.AsSpan();
        ref var start = ref MemoryMarshal.GetReference(animations);
        ref var end = ref Unsafe.Add(ref start, animations.Length);

        for (; Unsafe.IsAddressLessThan(ref start, ref end); start = ref Unsafe.Add(ref start, 1))
            if (start is TAnimation a)
                action(a);
    }

    public void Inlined<T>()
        where T : struct, Enum
    {
        var animations = _animations.AsSpan();
        ref var start = ref MemoryMarshal.GetReference(animations);
        ref var end = ref Unsafe.Add(ref start, animations.Length);

        for (; Unsafe.IsAddressLessThan(ref start, ref end); start = ref Unsafe.Add(ref start, 1))
            if (start is Animation<T> a)
                a.Change();
    }
}
```
```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:00.39]     InlineMethod.Tests.ModuleWeaverTests.InlineDelegate3 [FAIL]
  Failed InlineMethod.Tests.ModuleWeaverTests.InlineDelegate3 [2 ms]
  Error Message:
   Assert.True() Failure
Expected: True
Actual:   False
  Stack Trace:
     at InlineMethod.Tests.ModuleWeaverTests.CheckSimpleClass(TypeDefinition type, Boolean isEndsWith) in /media/f/RiderProjects/Quarantine/InlineMethod.Fody/src/InlineMethod.Tests/ModuleWeaverTests.cs:line 96
   at InlineMethod.Tests.ModuleWeaverTests.InlineDelegate3() in /media/f/RiderProjects/Quarantine/InlineMethod.Fody/src/InlineMethod.Tests/ModuleWeaverTests.cs:line 249
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
  Standard Output Messages:
 Method System.Void InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1::Caller()
 IL_0000: ldarg.0
 IL_0001: ldfld InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>[] InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1<TAnimation>::_animations
 IL_0006: call System.Span`1<!!0> System.MemoryExtensions::AsSpan<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0[])
 IL_000b: stloc.0
 IL_000c: ldloc.0
 IL_000d: call !!0& System.Runtime.InteropServices.MemoryMarshal::GetReference<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(System.Span`1<!!0>)
 IL_0012: stloc.1
 IL_0013: ldloc.1
 IL_0014: ldloca.s V_0
 IL_0016: call System.Int32 System.Span`1<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>::get_Length()
 IL_001b: call !!0& System.Runtime.CompilerServices.Unsafe::Add<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0&,System.Int32)
 IL_0020: stloc.2
 IL_0021: br.s IL_004b
 IL_0023: ldloc.1
 IL_0024: ldind.ref
 IL_0025: stloc.s V_4
 IL_0027: ldloc.s V_4
 IL_0029: isinst InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,T>
 IL_002e: brfalse.s IL_0043
 IL_0030: ldloc.s V_4
 IL_0032: isinst InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,T>
 IL_0037: unbox.any InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,T>
 IL_003c: stloc.3
 IL_003d: ldloc.3
 IL_003e: callvirt System.Void InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,T>>::Change()
 IL_0043: ldloc.1
 IL_0044: ldc.i4.1
 IL_0045: call !!0& System.Runtime.CompilerServices.Unsafe::Add<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0&,System.Int32)
 IL_004a: stloc.1
 IL_004b: ldloc.1
 IL_004c: ldloc.2
 IL_004d: call System.Boolean System.Runtime.CompilerServices.Unsafe::IsAddressLessThan<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0&,!!0&)
 IL_0052: brtrue.s IL_0023
 IL_0054: ret
 Method System.Void InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1::Inlined()
 IL_0000: ldarg.0
 IL_0001: ldfld InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>[] InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1<TAnimation>::_animations
 IL_0006: call System.Span`1<!!0> System.MemoryExtensions::AsSpan<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0[])
 IL_000b: stloc.0
 IL_000c: ldloc.0
 IL_000d: call !!0& System.Runtime.InteropServices.MemoryMarshal::GetReference<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(System.Span`1<!!0>)
 IL_0012: stloc.1
 IL_0013: ldloc.1
 IL_0014: ldloca.s V_0
 IL_0016: call System.Int32 System.Span`1<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>::get_Length()
 IL_001b: call !!0& System.Runtime.CompilerServices.Unsafe::Add<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0&,System.Int32)
 IL_0020: stloc.2
 IL_0021: br.s IL_003c
 IL_0023: ldloc.1
 IL_0024: ldind.ref
 IL_0025: isinst InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,T>
 IL_002a: stloc.3
 IL_002b: ldloc.3
 IL_002c: brfalse.s IL_0034
 IL_002e: ldloc.3
 IL_002f: callvirt System.Void InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,T>::Change()
 IL_0034: ldloc.1
 IL_0035: ldc.i4.1
 IL_0036: call !!0& System.Runtime.CompilerServices.Unsafe::Add<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0&,System.Int32)
 IL_003b: stloc.1
 IL_003c: ldloc.1
 IL_003d: ldloc.2
 IL_003e: call System.Boolean System.Runtime.CompilerServices.Unsafe::IsAddressLessThan<InlineMethod.Tests.AssemblyToProcess.InlineDelegate3`1/Animation`1<TAnimation,TAnimation>>(!!0&,!!0&)
 IL_0043: brtrue.s IL_0023
 IL_0045: ret



Failed!  - Failed:     1, Passed:    23, Skipped:     0, Total:    24, Duration: 49 ms - InlineMethod.Tests.dll (net8.0)
```

**Comment by oleg-st** [comment]
Is the code incorrect or just not the same as the manually inlined code?

**Comment by Emik03** [comment]
It produces the wrong IL. It effectively did this:
```cs
var animations = _animations.AsSpan();
ref var start = ref MemoryMarshal.GetReference(animations);
ref var end = ref Unsafe.Add(ref start, animations.Length);

for (; Unsafe.IsAddressLessThan(ref start, ref end); start = ref Unsafe.Add(ref start, 1))
    if (start is Animation<T>)
        ((Animation<Animation<T>>)start).Change(); // wrong cast!
```

**Comment by Emik03** [comment]
You can see the unexpected nested generic call being used in actual's `IL_003e`.

**Comment by oleg-st** [comment]
I think it's fixed in 0.8.2

**Comment by Emik03** [comment]
Awesome, it works for all of my current use cases, thank you once more for the hard work! I'll open a separate issue if I come across any bugs in the future.
